GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/Shadow/W3DProjectedShadow.cpp

```diff
@@ -1601,7 +1601,7 @@ Shadow* W3DProjectedShadowManager::addDecal(RenderObjClass *robj, Shadow::Shadow
 	//can be always perpendicular to model's z-axis or projected
 	//onto world geometry.
 	nameLen=strlen(shadowInfo->m_ShadowName);
-	strncpy(texture_name,shadowInfo->m_ShadowName,nameLen);
+	strlcpy(texture_name,shadowInfo->m_ShadowName,nameLen + 1);
```

Comment:
Essentially this is a strcpy now.

Replace

```cpp
nameLen=strlen(shadowInfo->m_ShadowName);
strlcpy(texture_name,shadowInfo->m_ShadowName,nameLen + 1);
strcpy(texture_name+nameLen,".tga");
```

with

```cpp
strcpy(texture_name, shadowInfo->m_ShadowName);
strlcpy(texture_name, ".tga", ARRAY_SIZE(texture_name));
```

And for robustness, declare `texture_name` like so:

```cpp
Char texture_name[ARRAY_SIZE(shadowInfo->m_ShadowName)];
```

Several times.